src/persist-client/src/read.rs

```diff
@@ -579,89 +707,51 @@ where
-            return Err(InvalidUsage::BatchNotFromThisShard {
-                batch_shard: batch.shard_id,
-                handle_shard: self.machine.shard_id(),
-            });
-        }
-
-        let mut updates = Vec::new();
-        for key in batch.batch.keys.iter() {
-            self.maybe_heartbeat_reader().await;
-            fetch_batch_part(
-                &batch.shard_id,
-                self.blob.as_ref(),
-                &self.metrics,
-                &key,
-                &batch.batch.desc.clone(),
-                |k, v, mut t, d| {
-                    match &batch.reader_metadata {
-                        HollowBatchReaderMetadata::Listen {
-                            as_of,
-                            until,
-                            since: _,
-                        } => {
-                            // This time is covered by a snapshot
-                            if !as_of.less_than(&t) {
-                                return;
-                            }
-
-                            // Because of compaction, the next batch we get might also
-                            // contain updates we've already emitted. For example, we
-                            // emitted `[1, 2)` and then compaction combined that batch
-                            // with a `[2, 3)` batch into a new `[1, 3)` batch. If this
-                            // happens, we just need to filter out anything < the
-                            // frontier. This frontier was the upper of the last batch
-                            // (and thus exclusive) so for the == case, we still emit.
-                            if !until.less_equal(&t) {
-                                return;
-                            }
-                        }
-                        HollowBatchReaderMetadata::Snapshot { as_of } => {
-                            // This time is covered by a listen
-                            if as_of.less_than(&t) {
-                                return;
-                            }
-                            t.advance_by(as_of.borrow())
-                        }
-                    }
-
-                    let k = self.metrics.codecs.key.decode(|| K::decode(k));
-                    let v = self.metrics.codecs.val.decode(|| V::decode(v));
-                    let d = D::decode(d);
-                    updates.push(((k, v), t, d));
-                },
-            )
-            .await;
+    /// Tracks that the `ReadHandle`'s machine's current `SeqNo` is being
+    /// "leased out" to a `ReaderEncirchedHollowBatch`, and cannot be garbage
+    /// collected until its lease has been dropped (`Self::drop_seqno_lease`).
+    pub fn lease_seqno(&mut self) -> SeqNo {
+        let lease = self.machine.seqno();
+        match self.leased_seqnos.last_mut() {
+            Some((seqno, leases)) if seqno == &lease => *leases += 1,
+            _ => {
+                // Sorted invariant maintained
```

Comment:
does this need to be a sorted Vec for performance/efficiency? seems like a HashMap might be more straightforward?

Reply:
I don't know what the expected throughput is, so erred on efficiency's side. We need to be able to find the minimum element while doing arbitrary deletes. If scanning all leases is fine, certainly `std` has better things on offer.

Reply:
I agree that a `HashMap` or sth simpler seems fine. We pay the cost when downgrading the since, and we don't do that super often.

Reply:
realized as I was falling asleep last night that we just want a std::collections::BTreeMap :)

Reply:
@danhhz Yeah--that's what I have locally!

Reply:
👌

Reply:
This is done in latest push